src/Crypto/Hash/HyperlaneNatives.hs

```diff
@@ -0,0 +1,340 @@
+  , hmMessageBody :: ByteString -- variable
+  }
+  deriving stock (Eq, Show)
+
+data TokenMessageERC20 = TokenMessageERC20
+  { tmAmount :: Word256 -- uint256
+  , tmChainId :: Word16 -- uint16
+  , tmRecipient :: ByteString -- variable
+  }
+  deriving stock (Eq, Show)
+
+----------------------------------------------
+--    Hyperlane Message Binary Encoding     --
+----------------------------------------------
+
+packHyperlaneMessage :: HyperlaneMessage -> Builder
+packHyperlaneMessage (HyperlaneMessage{..}) =
+  BB.word8 hmVersion
+  <> BB.word32BE hmNonce
+  <> BB.word32BE hmOriginDomain
+  <> BB.byteString hmSender
+  <> BB.word32BE hmDestinationDomain
+  <> BB.byteString hmRecipient
+  <> BB.byteString hmMessageBody
+
+-- The TokenMessage contains a recipient (text) and an amount (word-256).
+-- A schematic of the message format:
+-- 0000000000000000000000000000000000000000000000008ac7230489e80000 # amount = 10000000000000000000
+-- 0000 # chainId = 0
+-- 7B2270726564223A20226B6579732D616C6C222C20226B657973223A205B2264 # {"pred": "keys-all", "keys": ["da1a339bd82d2c2e9180626a00dc043275deb3ababb27b5738abf6b9dcee8db6"]}
+-- 6131613333396264383264326332653931383036323661303064633034333237
+-- 3564656233616261626232376235373338616266366239646365653864623622
+-- 5D7D
+--
+-- Note that we use abi.encodePacked here and the data is being concatenated without structure.
+-- For more details see https://docs.soliditylang.org/en/latest/abi-spec.html#non-standard-packed-mode
+packTokenMessageERC20 :: TokenMessageERC20 -> Builder
+packTokenMessageERC20 t =
+  word256BE (tmAmount t)
+  <> BB.word16BE (tmChainId t)
+  <> BB.byteString (tmRecipient t)
+
+unpackTokenMessageERC20 :: Get TokenMessageERC20
+unpackTokenMessageERC20 = do
+  tmAmount <- getWord256BE
+  tmChainId <- Bin.getWord16be
+  tmRecipient <- BL.toStrict <$> Bin.getRemainingLazyByteString
+
+  pure $ TokenMessageERC20 {..}
+
+----------------------------------------------
+--        Hyperlane Message Hashing         --
+----------------------------------------------
+
+getHyperlaneMessageId :: HyperlaneMessage -> Text
+getHyperlaneMessageId =
+  Text.decodeUtf8
+  . encodeBase64UrlUnpadded
+  . keccak256Hash
+  . BL.toStrict
+  . BB.toLazyByteString
+  . packHyperlaneMessage
+
+keccak256Hash :: ByteString -> ByteString
+keccak256Hash = BSS.fromShort . _getBytesN . _getKeccak256Hash . keccak256
+
+decodeBase64 :: FieldKey -> Text -> Either HyperlaneError ByteString
+decodeBase64 key s =
+  first (const (HyperlaneErrorInvalidBase64 key)) $ decodeBase64UrlUnpadded $ Text.encodeUtf8 s
+
+decodeBase64AndValidate :: FieldKey -> Int -> Text -> Either HyperlaneError ByteString
+decodeBase64AndValidate key expected s = do
+  decoded <- decodeBase64 key s
+
+  unless (BS.length decoded == expected) $
+    throwError $ HyperlaneErrorIncorrectSize key expected (BS.length decoded)
+
+  return decoded
+
+parseChainId :: forall a. (a ~ Word16) => Text -> Either HyperlaneError a
```

Comment:
if this has to be a `Word16`, why is this an equality constraint?

Reply:
refactoring purposes with show, i just wanted to write `a` everywhere. but you're right it's probably not worth it.